internal/manage/devices/download.go

```diff
@@ -0,0 +1,337 @@
+package devices
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/csv"
+	"fmt"
+	"iter"
+	"maps"
+	"net/http"
+	"net/url"
+	"slices"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protopath"
+	"google.golang.org/protobuf/reflect/protorange"
+	"google.golang.org/protobuf/reflect/protoreflect"
+
+	"github.com/smart-core-os/sc-api/go/traits"
+	"github.com/smart-core-os/sc-golang/pkg/resource"
+	"github.com/smart-core-os/sc-golang/pkg/trait"
+	"github.com/vanti-dev/sc-bos/pkg/gen"
+)
+
+//go:generate protomod protoc -- -I . -I ../../../proto --go_out=paths=source_relative:. download.proto
+
+func (s *Server) GetDownloadDevicesUrl(_ context.Context, request *gen.GetDownloadDevicesUrlRequest) (*gen.DownloadDevicesUrl, error) {
+	// validate
+	switch request.MediaType {
+	case "":
+		request.MediaType = "text/csv"
+	case "text/csv": // supported
+	default:
+		return nil, status.Errorf(codes.InvalidArgument, "unsupported media type %q", request.MediaType)
+	}
+
+	tokenData := &DownloadToken{
+		Request: request,
+	}
+	tokenStr, err := encodeDownloadToken(tokenData)
+	if err != nil {
+		return nil, err
+	}
+	u := s.downloadUrlBase
+	if err := s.encodeDownloadUrlToken(&u, tokenStr); err != nil {
```

Comment:
These encode/decode functions have very similar names and it's quite confusing. In particular `encodeDownloadUrlToken` which doesn't encode a token like I first thought based on the name, it encodes a URL based on a token which is already encoded. 

Could we merge these two steps? After all, all we really care about is being able to turn a `DownloadToken` into a URL and back again.

Reply:
I've changed the names to `read/writeDownloadToken` and `downloadTokenFrom/ToString` which I think helps. It makes sense to me to keep them separate to allow easier flow in the error cases.